src/mds/Server.cc

```diff
@@ -5089,6 +5126,64 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
       cur->add_waiter(CInode::WAIT_TRUNC, new C_MDS_RetryRequest(mdcache, mdr));
       return;
     }
+
+    if (truncating_smaller && req->get_data().length()) {
+      struct ceph_fscrypt_last_block_header header;
+      memset(&header, 0, sizeof(header));
+      auto bl = req->get_data().cbegin();
+      bufferlist *data = new bufferlist();
+      DECODE_START(1, bl);
+      decode(header.objver, bl);
+      decode(header.file_offset, bl);
+      decode(header.block_size, bl);
+      DECODE_FINISH(bl);
+
+      dout(20) << __func__ << " mdr->objver:" << mdr->objver
+               << " mdr->retry:" << mdr->retry
+               << " header.objver: " << header.objver
+               << " header.file_offset: " << header.file_offset
+               << " header.block_size: " << header.block_size
+               << dendl;
+
+      if (!header.objver) {
+        dout(20) << __func__ << " The last block is located in a file hole!"
+		 << dendl;
```

Comment:
Well, it was in a hole when the client did the read, but it might not be a hole anymore by this point in time. I think if you get a 0 objver from the client, then the operation needs to be gated on whether the backing object exists when you go to do the truncate.

Reply:
So strange, in many PRs I couldn't receive any notification about the new comments for recent weeks.

Yeah, make sense and I will fix it.